[addons] add first gui control addon class for buttons

AlwinEsch opened the pull request:
<!--- Provide a general summary of your change in the Title above -->

## Description

This add the first class on new addon interface to edit a button
control.

Hold on this alone to see better how the controls becomes used
on new addon system.

On next request are all controls together.

Here the smallest example to edit a control on new way:
```cpp
// this = The addon window class
// 33 = The on skin.xml used control id
kodi::gui::control::CButton(this, 33).SetVisible(false);
```

<!--- Describe your change in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

ksooo commented on xbmc/addons/interfaces/GUI/Window.cpp:
```diff
@@ -820,6 +823,25 @@ void Interface_GUIWindow::mark_dirty_region(void* kodiBase, void* handle)
+ * To get control functions
+ */
+//@{
+void* Interface_GUIWindow::getcontrol_button(void* kodiBase, void* handle, int control_id)
```
getcontrol_button => get_control_button? (missing underscore between "get" and "control")

ksooo commented on xbmc/addons/interfaces/GUI/Window.h:
```diff
@@ -104,6 +104,9 @@ namespace ADDON
     /* Various functions */
     static void mark_dirty_region(void* kodiBase, void* handle);
+
+    /* To get control functions */
```
"GUI control access functions"

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/ControlButton.h:
```diff
@@ -0,0 +1,175 @@
+  /// @note The call of the control is  only  possible  from  the  corresponding
+  /// window as its class and identification number is required.
+  ///
+  class CControlButton : public CAddonGUIControlBase
```
I would not use "CControl" prefix for all classes modelling GUI controls. This is what namespaces are for. 

I would use just "CButton" here. For me it is quite obvious  that a class contained in namespace "kodi::gui" named "CButton" is a GUI control. Even better, you could put all GUI controls in namespace kodi::gui::control.

AlwinEsch replied:
For me not so good to have it as "CButton" alone. In this case should also dialogs renamed and in own namespace.

Further want I hold it a bit equal with the Python where it is named "ControlButton"

ksooo replied:
You are the boss. At the end it is just names.

ksooo replied:
I think we should use the features of the languages we support. And C++ has namespaces, a concept which fits very nicely here. The fact that it is (currently) named the same for Python is not a very strong point here. And yes, dialogs should go to kodi::gui::dialogs, then.

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/definitions.h:
```diff
@@ -213,11 +223,15 @@ typedef struct AddonToKodiFuncTable_kodi_gui_window
   /* Various functions */
   void (*mark_dirty_region)(void* kodiBase, void* handle);
+
+  /* To get control functions */
```
"GUI control access functions"

ksooo reviewed (changes requested):
As always, some comments and suggestions.

AlwinEsch committed:
[cmake][addons] prevent the use of *.txt files

This prevent the use of text files from addon include dir during install
and define by cmake.

ksooo commented:
Very nice. Feel free to add another commit to change dialogs namespace (as we discussed on Slack).

AlwinEsch commented:
Currently on work of them :)

AlwinEsch committed:
[addons] add first gui control addon class for buttons

This add the first class on new addon interface to edit a button
control.

Hold on this alone to see better how the controls becomes used
on new addon system.

On next request are all controls together.

Here the smallest example to edit a control on new way:
```cpp
// this = The addon window class
// 33 = The on skin.xml used control id
kodi::gui::controls::CButton(this, 33).SetVisible(false);
```

AlwinEsch committed:
[addons] rename addon dialogs to use namespace kodi::gui::dialogs::..

Change to make the different gui parts more clear.

AlwinEsch committed:
[addons] fix CID 170221:  Null pointer dereferences

AlwinEsch committed:
[cmake] fix CMakeLists.txt include paths

ksooo reviewed (approved):
Good to merge, imo.

ksooo commented:
jenkins build this please

ksooo commented:
OSX-64 jenkins error is unrelated.
